mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp

```diff
@@ -2097,6 +2267,27 @@ class ArgMaxConverter : public OpRewritePattern<tosa::ArgMaxOp> {
               nestedLoc, predicate, newValue, oldValue);
           auto resultIndex = rewriter.create<arith::SelectOp>(
               nestedLoc, predicate, newIndex, oldIndex);
+
+          // Check if we need to materialize compare and select for the given
+          // NaN propagation mode.
+
+          // "PROPAGATE" matches the default NaN propagation mode of the arith
+          // dialect so no compare and select is required.
+          //
+          // In the case "IGNORE" we check if the current argument is NaN and
+          // select the old index and value otherwise take the updated index and
+          // value.
+          if (const auto nanMode = argmaxOp.getNanMode(); nanMode == "IGNORE") {
+            // Unordered comparison of NaN against itself will always return
+            // true.
+            Value isNaN = rewriter.create<arith::CmpFOp>(
```

Comment:
Can we refactor this with `materializeBinaryNanCheckIfRequired`?

Reply:
As above the NaN checking semantics in argmax don't match that of a binary or clamp since we have to propagate both the value and its index so we end up with one compare and two selects vs. two compares and two selects for binary ops and one compare and one select for clamp. Since these are all slightly different I chose to be explicit and handle the logic separately in each case.